utils/frame/remote-externalities/src/lib.rs

```diff
@@ -550,154 +522,83 @@ where
-					assert_eq!(thread_keys_chunk.len(), storage_data.len());
-
-					let mut batch_kv = Vec::with_capacity(thread_keys_chunk.len());
-					for (key, maybe_value) in thread_keys_chunk.iter().zip(storage_data) {
-						match maybe_value {
-							Some(data) => {
-								thread_key_chunk_values.push((key.clone(), data.clone()));
-								batch_kv.push((key.clone().0, data.0));
-							},
-							None => {
-								log::warn!(
-									target: LOG_TARGET,
-									"key {:?} had none corresponding value.",
-									&key
-								);
-								let data = StorageData(vec![]);
-								thread_key_chunk_values.push((key.clone(), data.clone()));
-								batch_kv.push((key.clone().0, data.0));
-							},
-						};
-					}
-
-					// Send this chunk to the main thread to start inserting.
-					thread_sender.unbounded_send(Message::Batch(batch_kv)).unwrap();
-					thread_key_values.extend(thread_key_chunk_values);
-				}
-
-				thread_sender.unbounded_send(Message::Terminated).unwrap();
-				thread_key_values
-			});
-
-			handles.push(handle);
-		}
+		// Check if we got responses for all submitted requests.
+		assert_eq!(keys.len(), storage_data.len());
 
-		// first, wait until all threads send a `Terminated` message, in the meantime populate
-		// `pending_ext`.
-		let mut terminated = 0usize;
-		let mut batch_failed = false;
-		let mut processed = 0usize;
-		loop {
-			match rx.next().await.unwrap() {
-				Message::Batch(kvs) => {
-					let kvs = kvs
-						.into_iter()
-						.filter(|(k, _)| !is_default_child_storage_key(k))
-						.collect::<Vec<_>>();
-					processed += kvs.len();
-					pending_ext.batch_insert(kvs);
-					log::info!(
-						target: LOG_TARGET,
-						"inserting keys progress = {:.0}% [{} / {}]",
-						(processed as f32 / keys.len() as f32) * 100f32,
-						processed,
-						keys.len(),
-					);
-				},
-				Message::BatchFailed(error) => {
-					log::error!(target: LOG_TARGET, "Batch processing failed: {:?}", error);
-					batch_failed = true;
-					break
-				},
-				Message::Terminated => {
-					terminated += 1;
-					if terminated == handles.len() {
-						break
-					}
+		let key_values = keys
+			.iter()
+			.zip(storage_data)
+			.map(|(key, maybe_value)| match maybe_value {
+				Some(data) => (key.clone(), data),
+				None => {
+					log::warn!(target: LOG_TARGET, "key {:?} had none corresponding value.", &key);
+					let data = StorageData(vec![]);
+					(key.clone(), data)
 				},
-			}
-		}
-
-		// Ensure all threads finished execution before returning.
-		let keys_and_values =
-			handles.into_iter().flat_map(|h| h.join().unwrap()).collect::<Vec<_>>();
-
-		if batch_failed {
-			return Err("Batch failed.")
-		}
+			})
+			.collect::<Vec<_>>();
 
-		Ok(keys_and_values)
+		let mut sp = Spinner::with_timer(Spinners::Dots, "Inserting keys into DB...".into());
+		let start = Instant::now();
+		pending_ext.batch_insert(key_values.clone().into_iter().map(|(k, v)| (k.0, v.0)));
```

Comment:
Not sure how slow this is but as it actually writes to sync I/O it might be better use the blocking threadpool in tokio:

```rust
let kv = keys_and_values.clone();
tokio::spawn_blocking(move || {
   let mut sp = Spinner::with_timer(Spinners::Dots, "Inserting keys into DB...".into());
   let start = Instant::now();
   pending_ext.batch_insert(key_values.clone().into_iter().map(|(k, v)| (k.0, v.0)));
}).await;
Ok(key_values)
```

Reply:
Is this expected to yield performance improvement if nothing else is happening on the main thread while it's inserting? 

I think I need to wrap the `pending_ext` in an Arc and/or Mutex to get it working inside the closure, not sure if I should proceed with that refactor.

Reply:
you don't have to do it in this PR but worth investing whether it makes things faster or not.

Ideally, we should do any DB operations in async code it might fuckup the scheduler in tokio

Reply:
My bad you probably don't need to move the stuff into closure as you await for it ^^